launcher/modplatform/EnsureMetadataTask.cpp

```diff
@@ -0,0 +1,532 @@
+#include "net/NetJob.h"
+#include "tasks/MultipleOptionsTask.h"
+
+static ModPlatform::ProviderCapabilities ProviderCaps;
+
+static ModrinthAPI modrinth_api;
+static FlameAPI flame_api;
+
+EnsureMetadataTask::EnsureMetadataTask(Mod* mod, QDir dir, ModPlatform::Provider prov) : Task(nullptr), m_index_dir(dir), m_provider(prov)
+{
+    auto hash = getHash(mod);
+    if (hash.isEmpty())
+        emitFail(mod);
+    else
+        m_mods.insert(hash, mod);
+}
+
+EnsureMetadataTask::EnsureMetadataTask(QList<Mod*>& mods, QDir dir, ModPlatform::Provider prov)
+    : Task(nullptr), m_index_dir(dir), m_provider(prov)
+{
+    for (auto* mod : mods) {
+        if (!mod->valid()) {
+            emitFail(mod);
+            continue;
+        }
+
+        auto hash = getHash(mod);
+        if (hash.isEmpty()) {
+            emitFail(mod);
+            continue;
+        }
+
+        m_mods.insert(hash, mod);
+    }
+}
+
+QString EnsureMetadataTask::getHash(Mod* mod)
+{
+    /* Here we create a mapping hash -> mod, because we need that relationship to parse the API routes */
+    QByteArray jar_data;
+    try {
+        jar_data = FS::read(mod->fileinfo().absoluteFilePath());
+    } catch (FS::FileSystemException& e) {
+        qCritical() << QString("Failed to open / read JAR file of %1").arg(mod->name());
+        qCritical() << QString("Reason: ") << e.cause();
+
+        return {};
+    }
+
+    switch (m_provider) {
+        case ModPlatform::Provider::MODRINTH: {
+            auto hash_type = ProviderCaps.hashType(ModPlatform::Provider::MODRINTH).first();
+
+            return QString(ProviderCaps.hash(ModPlatform::Provider::MODRINTH, jar_data, hash_type).toHex());
+        }
+        case ModPlatform::Provider::FLAME: {
+            QByteArray jar_data_treated;
+            for (char c : jar_data) {
+                // CF-specific
+                if (!(c == 9 || c == 10 || c == 13 || c == 32))
+                    jar_data_treated.push_back(c);
+            }
+
+            return QString::number(MurmurHash2(jar_data_treated, jar_data_treated.length()));
+        }
+    }
+
+    return {};
+}
+
+bool EnsureMetadataTask::abort()
+{
+    // Prevent sending signals to a dead object
+    disconnect(this, 0, 0, 0);
+
+    if (m_current_task)
+        return m_current_task->abort();
+    return true;
+}
+
+void EnsureMetadataTask::executeTask()
+{
+    setStatus(tr("Checking if mods have metadata..."));
+
+    for (auto* mod : m_mods) {
+        if (!mod->valid())
+            continue;
+
+        // They already have the right metadata :o
+        if (mod->status() != ModStatus::NoMetadata && mod->metadata() && mod->metadata()->provider == m_provider) {
+            qDebug() << "Mod" << mod->name() << "already has metadata!";
+            emitReady(mod);
+            return;
+        }
+
+        // Folders don't have metadata
+        if (mod->type() == Mod::MOD_FOLDER) {
+            emitReady(mod);
+            return;
+        }
+    }
```

Comment:
fix : ACTUALLY support more than one mod

```suggestion
    for (auto* mod : m_mods) {
        if (!mod->valid())
            continue;

        // They already have the right metadata :o
        if (mod->status() != ModStatus::NoMetadata && mod->metadata() && mod->metadata()->provider == m_provider) {
            qDebug() << "Mod" << mod->name() << "already has metadata!";
            emitReady(mod);
            continue;
        }

        // Folders don't have metadata
        if (mod->type() == Mod::MOD_FOLDER) {
            emitReady(mod);
        }
    }
```